### DISCUSSION TITLE
wasm-ld memory usage AOT

### DISCUSSION THREAD

**Discussion by andrey-shavelev** [unoplatform/uno#9768] [Q&A]
Dear All, when compiling uno.wasm project with AOT, wasm-ld is used internally. One of the problems with it is that it is very demanding to memory. In my case, wasm-ld takes up to 27 gigabytes. And if a build agent does not have that amount (and my default BA doesn't have it) the process crashes with "out of memory". 
I tried to set <WasmShellExtraEmccFlags Include="-s MAXIMUM_MEMORY=8GB"/> or UNO_BOOTSTRAP_EMSCRIPTEN_MAXIMUM_MEMORY=8GB. However, these flags either do not affect the wasm-ld, or are not always passed to it. Eventually, the memory usage grows and the process crashes.

Could you advise me if there is a way to pass memory-limiting parameters to wasm-ld?

**Comment by jeromelaban** [comment]
wasm-ld is indeed a piece of the emscripten tooling that can take a significant amount of memory, particularly when using AOT without a [profile](https://platform.uno/docs/articles/external/uno.wasm.bootstrap/doc/runtime-execution-modes.html#profile-guided-aot). You can change the [linker optimization settings](https://github.com/unoplatform/Uno.Wasm.Bootstrap/blob/main/doc/features-linker-opts.md#emscripten-linker-optimizations-flags) to lower the optimization values.

**Reply by andrey-shavelev** [reply]
Thanks for the reply! 
Could this flag be used to limit memory usage?
https://lld.llvm.org/WebAssembly.html#cmdoption-max-memory

**Reply by jeromelaban** [reply]
Could be, but we can't pass parameters to wasm-ld from emscripten, last time I checked. Maybe it will change in the future (or a later version that we can use because .NET uses it).
